### PR TITLE
[WIP] Push notifications

### PR DESCRIPTION
**Pull request by tgxworld**

### CODE REVIEW

File: app/assets/javascripts/discourse/components/desktop-notification-config.js.es6 [hunk 1]
```diff
@@ -59,11 +90,28 @@ export default Ember.Component.extend({
...
       this.propertyDidChange('notificationsPermission');
+      this.send('unsubscribe');
     },
 
     turnon() {
       this.set('notificationsDisabled', '');
       this.propertyDidChange('notificationsPermission');
+    },
+    subscribe() {
+      const self = this;
```
Comment (lcpriest) [inline app/assets/javascripts/discourse/components/desktop-notification-config.js.es6]: Is this necessary if you are using the fat arrow below?

File: app/assets/javascripts/discourse/lib/push-notifications.js.es6 [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+export function unsubscribe(callback) {
+  if (!isPushNotificationsSupported()) return;
+
+  navigator.serviceWorker.ready.then(serviceWorkerRegistration => {
+    serviceWorkerRegistration.pushManager.getSubscription().then(subscription => {
+      if (subscription) {
+        subscription.unsubscribe().then((successful) => {
+          if (successful) Discourse.ajax('/push_notifications/unsubscribe', {
```
Comment (lcpriest) [inline app/assets/javascripts/discourse/lib/push-notifications.js.es6]: Is Discourse.ajax not `import`able?

File: app/controllers/push_notifications_controller.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    notifications = Notification.where(user_id: current_user.id, read: false)
+                                .visible
+                                .includes(:topic)
+                                .order(created_at: :desc)
+
+    base_url = Discourse.base_url
+    site_title = SiteSetting.title
+
+    info =
+      if notifications.count == 1
```
Comment (lcpriest) [inline app/controllers/push_notifications_controller.rb]: If you invert this conditional you can use `.many?`
http://api.rubyonrails.org/classes/Enumerable.html#method-i-many-3F

File: config/routes.rb [hunk 4]
```diff
@@ -572,6 +572,14 @@
     end
   end
 
+  resources :push_notifications do
+    collection do
+      post "subscribe"
```
Comment (lcpriest) [inline config/routes.rb]: These can be symbols right?

File: app/assets/javascripts/discourse/components/desktop-notification-config.js.es6 (hunk 1 above)
Comment (tgxworld) [inline app/assets/javascripts/discourse/components/desktop-notification-config.js.es6]: :+1:

File: app/assets/javascripts/discourse/lib/push-notifications.js.es6 (hunk 2 above)
Comment (tgxworld) [inline app/assets/javascripts/discourse/lib/push-notifications.js.es6]: Nope it isn't written as an es6 module yet.

File: app/controllers/push_notifications_controller.rb (hunk 3 above)
Comment (tgxworld) [inline app/controllers/push_notifications_controller.rb]: O cool! Didn't know about `.many?` :+1:

File: config/routes.rb (hunk 4 above)
Comment (tgxworld) [inline config/routes.rb]: Yup they can be.

**Comment by coding-horror** [comment]
this should also wait until 1.6 is beta

**Comment by SamSaffron** [comment]
will reopen (or create a plugin) when ready

**Pull request closed without merging**